File: backend/app/schemas/inspection.py
"""
Inspection Schemas - INSP-01/02
Matches frontend/src/types/inspection.ts
"""
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class ProcessFlowDTO(BaseModel):
    """Process flow"""
    flow_id: str
    site_id: str
    flow_name: str
    rev_no: int
    status: str  # 'DRAFT' | 'ACTIVE' | 'OBSOLETE'
    description: Optional[str]
    created_at: datetime
    created_by: Optional[str]
    updated_at: datetime
    updated_by: Optional[str]


class ProcessStepDTO(BaseModel):
    """Process step"""
    step_id: str
    flow_id: str
    seq_no: int
    step_name: str
    workcenter: Optional[str]
    machine_group: Optional[str]
    is_inspection_point: bool
    created_at: datetime
    created_by: Optional[str]
    updated_at: datetime
    updated_by: Optional[str]


class AIProcessDesignRequest(BaseModel):
    """AI process design request"""
    company: dict
    product: dict
    constraints: dict
    known_defects: List[str]
    customer_requirements: List[str]


class AIProcessDesignResponse(BaseModel):
    """AI process design response"""
    ai_id: str
    process_flow: List[dict]
    inspection_points: List[dict]
    assumptions: List[str]
    risks: List[str]
    confidence: float


class AICriteriaChecklistRequest(BaseModel):
    """AI criteria checklist request"""
    item: str
    step: str
    inspection_type: str
    defect_modes: List[str]
    acceptance_policy: dict


class AICriteriaChecklistResponse(BaseModel):
    """AI criteria checklist response"""
    ai_id: str
    criteria_table: List[dict]
    checklist: List[dict]


class InspectionRunDTO(BaseModel):
    """Inspection run"""
    run_id: str
    site_id: str
    lot_id: str
    plan_id: str
    step_id: str
    run_dt: datetime
    inspector_id: Optional[str]
    sample_n: int
    environment: dict = {}
    overall_judgement: str  # 'PASS' | 'FAIL' | 'HOLD'
    ai_summary_id: Optional[str]
    created_at: datetime
    created_by: Optional[str]
    updated_at: datetime
    updated_by: Optional[str]


class InspectionResultDTO(BaseModel):
    """Inspection result"""
    result_id: str
    run_id: str
    char_id: str
    spec_id: str
    sample_no: int
    value_num: Optional[float]
    value_text: Optional[str]
    value_enum: Optional[str]
    is_out_of_spec: bool
    defect_code_id: Optional[str]
    remark: Optional[str]
    created_at: datetime
    created_by: Optional[str]
    updated_at: datetime
    updated_by: Optional[str]


class BulkResultRequest(BaseModel):
    """Bulk result request"""
    results: List[dict]


class BulkResultResponse(BaseModel):
    """Bulk result response"""
    inserted: int
    oos_count: int


class InspectionJudgeResponse(BaseModel):
    """Inspection judgement response"""
    run_id: str
    overall_judgement: str  # 'PASS' | 'FAIL' | 'HOLD'
    oos_count: int

File: frontend/src/types/inspection.ts
/**
 * Inspection Types
 * INSP-01/02: 검사 프로세스 및 실행 API 타입 정의
 */

export interface ProcessFlowDTO {
  flow_id: string;
  site_id: string;
  flow_name: string;
  rev_no: number;
  status: 'DRAFT' | 'ACTIVE' | 'OBSOLETE';
  description: string | null;
  created_at: string;
  created_by: string | null;
  updated_at: string;
  updated_by: string | null;
}

export interface ProcessStepDTO {
  step_id: string;
  flow_id: string;
  seq_no: number;
  step_name: string;
  workcenter: string | null;
  machine_group: string | null;
  is_inspection_point: boolean;
  created_at: string;
  created_by: string | null;
  updated_at: string;
  updated_by: string | null;
}

export interface AIProcessDesignRequest {
  company: {
    industry: string;
    site: string;
  };
  product: {
    item_name: string;
    process: string;
  };
  constraints: {
    inspection_headcount: number;
    shift: number;
  };
  known_defects: string[];
  customer_requirements: string[];
}

export interface AIProcessDesignResponse {
  ai_id: string;
  process_flow: ProcessFlowDTO[];
  inspection_points: ProcessStepDTO[];
  assumptions: string[];
  risks: string[];
  confidence: number;
}

export interface AICriteriaChecklistRequest {
  item: string;
  step: string;
  inspection_type: string;
  defect_modes: string[];
  acceptance_policy: {
    A: string;
    B: string;
    C: string;
  };
}

export interface AICriteriaChecklistResponse {
  ai_id: string;
  criteria_table: CriteriaTable[];
  checklist: ChecklistItem[];
}

export interface CriteriaTable {
  category: string;
  check: string;
  method: string;
  accept: string;
  reject: string;
  action: string;
}

export interface ChecklistItem {
  no: number;
  question: string;
  type: string;
  photo_required: boolean;
}

export interface InspectionRunDTO {
  run_id: string;
  site_id: string;
  lot_id: string;
  plan_id: string;
  step_id: string;
  run_dt: string;
  inspector_id: string | null;
  sample_n: number;
  environment: Record<string, any>;
  overall_judgement: 'PASS' | 'FAIL' | 'HOLD';
  ai_summary_id: string | null;
  created_at: string;
  created_by: string | null;
  updated_at: string;
  updated_by: string | null;
}

export interface InspectionResultDTO {
  result_id: string;
  run_id: string;
  char_id: string;
  spec_id: string;
  sample_no: number;
  value_num: number | null;
  value_text: string | null;
  value_enum: string | null;
  is_out_of_spec: boolean;
  defect_code_id: string | null;
  remark: string | null;
  created_at: string;
  created_by: string | null;
  updated_at: string;
  updated_by: string | null;
}

export interface BulkResultRequest {
  results: Omit<InspectionResultDTO, 'result_id' | 'created_at' | 'created_by' | 'updated_at' | 'updated_by'>[];
}

export interface BulkResultResponse {
  inserted: number;
  oos_count: number;
}

export interface InspectionJudgeResponse {
  run_id: string;
  overall_judgement: 'PASS' | 'FAIL' | 'HOLD';
  oos_count: number;
}
